operator/pkg/ciliumendpointbatch/reconciler.go

```diff
@@ -0,0 +1,131 @@
+// Copyright 2021 Authors of Cilium
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package ciliumendpointbatch
+
+import (
+	"context"
+
+	"k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+
+	cilium_v2 "github.com/cilium/cilium/pkg/k8s/apis/cilium.io/v2alpha1"
+	clientset "github.com/cilium/cilium/pkg/k8s/client/clientset/versioned/typed/cilium.io/v2alpha1"
+	"github.com/sirupsen/logrus"
+)
+
+// reconciler is used to sync the current (i.e. desired) state of the CEBs in datastore into current state CEBs in the k8s-apiserver.
+// The source of truth is in local datastore, client reconcile with apiserver at regular intervals.
+type reconciler struct {
+	client     clientset.CiliumV2alpha1Interface
+	cebManager cebManager
+}
+
+// newReconciler creates and initializes a new reconciler.
+func newReconciler(client clientset.CiliumV2alpha1Interface, cebMgr cebManager) *reconciler {
+	return &reconciler{
+		client:     client,
+		cebManager: cebMgr,
+	}
+}
+
+// reconcileWithServer gets the list of CEBs to create, update and delete and syncs with the api-server.
+// On error, it returns the list of failed createCebs, updateCebs and deleteCebs to the caller.
+func (r *reconciler) reconcileWithServer(cebsToCreate, cebsToUpdate, cebsToDelete []string) (errCebCreate, errCebUpdate, errCebDelete []string) {
+	var err error
+	var retCeb, ceb *cilium_v2.CiliumEndpointBatch
+
+	// Loop through list of cebsToUpdate and update the CEB
+	for _, c := range cebsToUpdate {
+
+		// Before syncing with ApiServer, get list of removed CEPs
+		remCeps := r.cebManager.getRemovedCeps(c)
+		// Get the copy of Ceb from the cebManager
+		if ceb, err = r.cebManager.getCebFromCache(c); err != nil {
+			continue
+		}
+
+		if retCeb, err = r.client.CiliumEndpointBatches().Update(
+			context.TODO(), ceb, metav1.UpdateOptions{}); err != nil {
+
+			// If CEB is not found in server, instead of resync try re-create new CEB.
```

Comment:
I think this is a nuance in implementation. The intention is to let apiserver end up in a state as the aggregator (CEB-wise).

The chain of event that could lead to this is:
- some CEB X is being deleted (because all of the CEPs in it are deleted)
- synchronizer tries to sync cebDelete for X with apiserver
- b4 the above req returns, one of the CEPs in X is re-created
- aggregator would cache a cebUpdate for X
- cebDelete goes through and returns (now apiserver does not have X)
- next time when aggregator is trying to send update for X, there will be an expected NotFound error, so just switch to CREATE request, what matters is the end state of apiserver

Reply:
Makes sense. It is nuanced, maybe a comment here saying
"This might happen if we recreate a CEB that was previously empty and raced with the delete"

Reply:
This shouldn't happen, I had removed this case.

Reply:
Why wouldn't this happen?

Reply:
Couple of reasons.

1. CEB is marked for DELETE, only if all CEPs in the CEB are deleted and successfully synced with api-server. That means, if  'CEPs' are re-created, this would queued up in a new CEB.
2. `CEB` should be updated only if CEB is created successfully with api-server.